src/screens/Routes/onNavigationStateChange.android.js

```diff
@@ -0,0 +1,88 @@
+import analytics from '@segment/analytics-react-native';
+import { get } from 'lodash';
+import { StatusBar } from 'react-native';
+import { NavigationActions } from 'react-navigation';
+import { Navigation } from '../../navigation';
+import { sentryUtils } from '../../utils';
+import { ROUTES } from './routesNames';
+
+export function onNavigationStateChange(prevState, currentState) {
+  const { params, routeName } = Navigation.getActiveRoute(currentState);
+  const prevRouteName = Navigation.getActiveRouteName(prevState);
+  // native stack rn does not support onTransitionEnd and onTransitionStart
+  // Set focus manually on route changes
+  if (prevRouteName !== routeName) {
+    Navigation.handleAction(
+      NavigationActions.setParams({
+        key: routeName,
+        params: { focused: true },
+      })
+    );
+
+    Navigation.handleAction(
+      NavigationActions.setParams({
+        key: prevRouteName,
+        params: { focused: false },
+      })
+    );
+  }
+
+  if (
+    prevRouteName !== ROUTES.QR_SCANNER_SCREEN &&
+    routeName === ROUTES.QR_SCANNER_SCREEN
```

Comment:
While this works for the small amount of screens, the number of combinations growth rate is `~O(n * (n - (n - c)))` where `n` is a number of routes and `c` is an average number of allowed transitions from each route of n. In the worst case, it gives us *almost* `O(n^2)` which is quite tedious to cover with scenarios like "if we navigate from A to B, do X. If we navigate from A to C, do Y". 

In my opinion, a more appropriate approach would be to either use a state machine here or move this information to the routes themselves and use the current approach to set the StatsBar color. It'll be not only faster (speed doesn't really matter here), but significantly easier to maintain in the long run.

Reply:
We know it's not a good approach. It was just a quick workaround we've done once and then repeated again and again... and again.